packages/@aws-cdk/custom-resources/lib/provider-framework/runtime/framework.ts

```diff
@@ -0,0 +1,177 @@
+ * interact with the user-defined `onEvent` and `isComplete` handlers.
+ *
+ * This function will always succeed. If an error occurs
+ *
+ * @param cfnRequest The cloudformation custom resource event.
+ */
+async function onEvent(cfnRequest: AWSLambda.CloudFormationCustomResourceEvent) {
+  await failOnError(cfnRequest, async () => {
+    log('onEventHandler', cfnRequest);
+
+    cfnRequest.ResourceProperties = cfnRequest.ResourceProperties || { };
+
+    // validate incoming request
+    validateCfnRequest(cfnRequest);
+
+    const onEventResult = await invokeUserFunction(consts.USER_ON_EVENT_FUNCTION_ARN_ENV, cfnRequest) as OnEventResponse;
+    log('onEvent returned:', onEventResult);
+
+    // merge the request and the result from onEvent to form the complete resource event
+    // this also performs validation.
+    const resourceEvent = createResponseEvent(cfnRequest, onEventResult);
+    log('event:', onEventResult);
+
+    // determine if this is an async provider based on whether we have an isComplete handler defined.
+    // if it is not defined, then we are basically ready to return a positive response.
+    if (!process.env[consts.USER_IS_COMPLETE_FUNCTION_ARN_ENV]) {
+      await submitCloudFormationResponse('SUCCESS', resourceEvent);
+      return;
+    }
+
+    // ok, we are not complete, so kick off the waiter workflow
+    const waiter = {
+      stateMachineArn: getEnv(consts.WAITER_STATE_MACHINE_ARN_ENV),
+      name: resourceEvent.RequestId,
+      input: JSON.stringify(resourceEvent),
+    };
+
+    log('starting waiter', waiter);
+
+    // kick off waiter state machine
+    await startExecution(waiter);
+    return;
+  });
+}
+
+// invoked a few times until `complete` is true or until it times out.
+async function isComplete(event: AWSCDKAsyncCustomResource.IsCompleteRequest) {
+  return await failOnError(event, async () => {
+    log('isComplete', event);
+
+    const isCompleteResult = await invokeUserFunction(consts.USER_IS_COMPLETE_FUNCTION_ARN_ENV, event) as IsCompleteResponse;
+    log('user isComplete returned:', isCompleteResult);
+
+    // if we are not complete, reeturn false, and don't send a response back.
+    if (!isCompleteResult.IsComplete) {
+      if (isCompleteResult.Data && Object.keys(isCompleteResult.Data).length > 0) {
+        throw new Error(`"Data" is not allowed if "IsComplete" is "False"`);
+      }
+
+      throw new Retry(JSON.stringify(event));
+    }
+
+    const response = {
+      ...event,
+      Data: {
+        ...event.Data,
+        ...isCompleteResult.Data
+      }
+    };
+
+    await submitCloudFormationResponse('SUCCESS', response);
+  });
+}
+
+// invoked when completion retries are exhaused.
+async function onTimeout(timeoutEvent: any) {
+  log('timeoutHandler', timeoutEvent);
+
+  const isCompleteRequest = JSON.parse(JSON.parse(timeoutEvent.Cause).errorMessage) as AWSCDKAsyncCustomResource.IsCompleteRequest;
+  await submitCloudFormationResponse('FAILED', isCompleteRequest, {
```

Comment:
Nit: no need to `await`, can just `return` this. Function doesn't even need to be marked `async`.